sklearn/semi_supervised/self_training.py

```diff
@@ -0,0 +1,250 @@
+        a sample has iteration -1, the sample was not labeled in any iteration.
+
+    n_iter_ : int
+        The amount of iterations the classifier during fitting.
+
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn import datasets
+    >>> from sklearn.semi_supervised import SelfTrainingClassifier
+    >>> from sklearn.svm import SVC
+    >>> svc = SVC(probability=True, gamma="auto")
+    >>> self_training_model = SelfTrainingClassifier(svc)
+    >>> iris = datasets.load_iris()
+    >>> rng = np.random.RandomState(42)
+    >>> random_unlabeled_points = rng.rand(len(iris.target)) < 0.3
+    >>> labels = np.copy(iris.target)
+    >>> labels[random_unlabeled_points] = -1
+    >>> self_training_model.fit(iris.data, labels)
+    ... # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    SelfTrainingClassifier(...)
+
+    References
+    ----------
+    David Yarowsky. 1995. Unsupervised word sense disambiguation rivaling
+    supervised methods. In Proceedings of the 33rd annual meeting on
+    Association for Computational Linguistics (ACL '95). Association for
+    Computational Linguistics, Stroudsburg, PA, USA, 189-196. DOI:
+    https://doi.org/10.3115/981658.981684
+    """
+    _estimator_type = "classifier"
+
+    def __init__(self,
+                 base_estimator,
+                 threshold=0.75,
+                 max_iter=100):
+        self.base_estimator = base_estimator
+        self.threshold = threshold
+        self.max_iter = max_iter
+
+    def fit(self, X, y):
+        """
+        Fits this ``SelfTrainingClassifier`` to dataset, using the
+        base_estimator passed. The fitted base_estimator is stored as
+        ``self.base_estimator_``.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            array representing the data
+
+        y : array-like, shape = (n_samples,)
+            array representing the labels. Unlabeled samples should have the
+            label -1.
+
+        Returns
+        -------
+        self: returns an instance of self.
+        """
+        X, y = check_X_y(X, y)
+        _validate_estimator(self.base_estimator)
+        self.base_estimator_ = clone(self.base_estimator)
+
+        if not 0 <= self.max_iter:
+            raise ValueError("max_iter must be >= 0")
+
+        if not 0 <= self.threshold < 1:
+            raise ValueError("threshold must be in [0,1)")
+
+        if y.dtype.kind == 'U' or y.dtype.kind == 'S':
+            has_label = y != '-1'
+        else:
+            has_label = y != -1
+
+        if np.logical_and.reduce(has_label):
+            warnings.warn(RuntimeWarning("y contains no unlabeled samples"))
+
+        self.y_labels_ = np.copy(y)
+        self.y_labeled_iter_ = np.full_like(y, -1)
+        self.y_labeled_iter_[has_label] = 0
+
+        self.n_iter_ = 0
+        while np.sum(has_label) < len(X) and self.n_iter_ < self.max_iter:
```

Comment:
`while not np.all(has_label) and self.n_iter_ < self.max_iter:`

In general, avoid using `len()` on numpy arrays. Use `array.shape[x]` or `array.size`.

Reply:
Out of own interest, why do you recommend avoiding `len()`?

Reply:
Mostly because using `shape` is the 'accepted' way. I think that's because it's not obvious at first sight what `len` would do on a numpy array: it could be the length of the first dimension or the number of entries in the array. At least shape is non-ambiguous.